File: src/Application Specific/Classes/Entity Collections/Jobs/EngineerVisit/EngineerVisitAssetWorkSheet/EngineerVisitAssetWorkSheetSQL.vb
Imports System.Data.SqlClient

Namespace Entity

    Namespace EngineerVisitAssetWorkSheets

        Public Class EngineerVisitAssetWorkSheetSQL
            Private _database As Sys.Database

            Public Sub New(ByVal database As Sys.Database)
                _database = database
            End Sub

#Region "Functions"

            Public Sub Delete(ByVal EngineerVisitAssetWorkSheetID As Integer)
                _database.ClearParameter()
                _database.AddParameter("@EngineerVisitAssetWorkSheetID", EngineerVisitAssetWorkSheetID, True)
                _database.ExecuteSP_NO_Return("EngineerVisitAssetWorkSheet_Delete")
            End Sub

            Public Function [EngineerVisitAssetWorkSheet_Get](ByVal EngineerVisitAssetWorkSheetID As Integer) As EngineerVisitAssetWorkSheet
                _database.ClearParameter()
                _database.AddParameter("@EngineerVisitAssetWorkSheetID", EngineerVisitAssetWorkSheetID)

                'Get the datatable from the database store in dt
                Dim dt As DataTable = _database.ExecuteSP_DataTable("EngineerVisitAssetWorkSheet_Get")

                If Not dt Is Nothing Then
                    If dt.Rows.Count > 0 Then
                        Dim oEngineerVisitAssetWorkSheet As New EngineerVisitAssetWorkSheet
                        With oEngineerVisitAssetWorkSheet
                            .IgnoreExceptionsOnSetMethods = True
                            .SetEngineerVisitAssetWorkSheetID = dt.Rows(0).Item("EngineerVisitAssetWorkSheetID")
                            .SetEngineerVisitID = dt.Rows(0).Item("EngineerVisitID")
                            .SetAssetID = dt.Rows(0).Item("AssetID")
                            .SetFlueTerminationSatisfactoryID = dt.Rows(0).Item("FlueTerminationSatisfactoryID")
                            .SetFlueFlowTestID = dt.Rows(0).Item("FlueFlowTestID")
                            .SetSpillageTestID = dt.Rows(0).Item("SpillageTestID")
                            .SetVentilationProvisionSatisfactoryID = dt.Rows(0).Item("VentilationProvisionSatisfactoryID")
                            .SetSafetyDeviceOperationID = dt.Rows(0).Item("SafetyDeviceOperationID")
                            .SetDHWFlowRate = dt.Rows(0).Item("DHWFlowRate")
                            .SetColdWaterTemp = dt.Rows(0).Item("ColdWaterTemp")
                            .SetDHWTemp = dt.Rows(0).Item("DHWTemp")
                            .SetInletStaticPressure = dt.Rows(0).Item("InletStaticPressure")
                            .SetInletWorkingPressure = dt.Rows(0).Item("InletWorkingPressure")
                            .SetMinBurnerPressure = dt.Rows(0).Item("MinBurnerPressure")
                            .SetMaxBurnerPressure = dt.Rows(0).Item("MaxBurnerPressure")
                            .SetCO2 = dt.Rows(0).Item("CO2")
                            .SetCO2CORatio = dt.Rows(0).Item("CO2CORatio")
                            .SetLandlordsApplianceID = dt.Rows(0).Item("LandlordsApplianceID")
                            .SetApplianceServiceOrInspectedID = dt.Rows(0).Item("ApplianceServiceOrInspectedID")
                            .SetApplianceTestedID = dt.Rows(0).Item("ApplianceTestedID")
                            .SetApplianceSafeID = dt.Rows(0).Item("ApplianceSafeID")
                            .SetVisualConditionOfFlueSatisfactoryID = dt.Rows(0).Item("VisualConditionOfFlueSatisfactoryID")
                            .SetDeleted = dt.Rows(0).Item("Deleted")
                            .SetNozzle = dt.Rows(0).Item("Nozzle")
                            .SetReading = dt.Rows(0).Item("ReadingType")
                            .SetTankID = dt.Rows(0).Item("TankID")
                            .SetBMake = dt.Rows(0).Item("BurMake")
                            .SetBModel = dt.Rows(0).Item("BurModel")
                            .SetSweepOutcomeID = dt.Rows(0).Item("SweepOutcomeID")
                            .SetOverallID = dt.Rows(0).Item("OverallID")
                            .SetDischargeID = dt.Rows(0).Item("DischargeID")
                        End With
                        oEngineerVisitAssetWorkSheet.Exists = True
                        Return oEngineerVisitAssetWorkSheet
                    Else
                        Return Nothing
                    End If
                Else
                    Return Nothing
                End If
            End Function

            Public Function [EngineerVisitAssetWorkSheet_GetAll]() As DataView
                _database.ClearParameter()
                Dim dt As DataTable = _database.ExecuteSP_DataTable("EngineerVisitAssetWorkSheet_GetAll")
                dt.TableName = Entity.Sys.Enums.TableNames.tblEngineerVisitAssetWorkSheet.ToString
                Return New DataView(dt)
            End Function

            Public Function EngineerVisitAssetWorkSheet_GetForVisit(ByVal EngineerVisitID As Integer, Optional ByVal Oil As Integer = -1) As DataView
                _database.ClearParameter()
                _database.AddParameter("@EngineerVisitID", EngineerVisitID, True)
                _database.AddParameter("@Fuel", Oil, True)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("EngineerVisitAssetWorkSheet_GetForVisit")
                dt.TableName = Entity.Sys.Enums.TableNames.tblEngineerVisitAssetWorkSheet.ToString
                Return New DataView(dt)
            End Function

            Public Function Get_Friendly(ByVal engineerVisitAssetWorkSheetId As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@EngineerVisitAssetWorkSheetID", engineerVisitAssetWorkSheetId, True)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("EngineerVisitAssetWorkSheet_Get_Friendly")
                dt.TableName = Entity.Sys.Enums.TableNames.tblEngineerVisitAssetWorkSheet.ToString
                Return New DataView(dt)
            End Function

            Public Function [Insert](ByVal oEngineerVisitAssetWorkSheet As EngineerVisitAssetWorkSheet) As EngineerVisitAssetWorkSheet
                _database.ClearParameter()
                AddEngineerVisitAssetWorkSheetParametersToCommand(oEngineerVisitAssetWorkSheet)

                oEngineerVisitAssetWorkSheet.SetEngineerVisitAssetWorkSheetID = Entity.Sys.Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("EngineerVisitAssetWorkSheet_Insert"))
                oEngineerVisitAssetWorkSheet.Exists = True
                Return oEngineerVisitAssetWorkSheet
            End Function

            Public Function [EngineerVisitAssetWorkSheet_Search](ByVal criteria As String) As DataView
                _database.ClearParameter()
                _database.AddParameter("@Criteria", criteria, True)

                Dim dt As DataTable = _database.ExecuteSP_DataTable("EngineerVisitAssetWorkSheet_Search")
                dt.TableName = Entity.Sys.Enums.TableNames.tblEngineerVisitAssetWorkSheet.ToString
                Return New DataView(dt)
            End Function

            Public Sub [Update](ByVal oEngineerVisitAssetWorkSheet As EngineerVisitAssetWorkSheet)
                _database.ClearParameter()
                _database.AddParameter("@EngineerVisitAssetWorkSheetID", oEngineerVisitAssetWorkSheet.EngineerVisitAssetWorkSheetID, True)
                AddEngineerVisitAssetWorkSheetParametersToCommand(oEngineerVisitAssetWorkSheet)
                _database.ExecuteSP_NO_Return("EngineerVisitAssetWorkSheet_Update")
            End Sub

            Private Sub AddEngineerVisitAssetWorkSheetParametersToCommand(ByRef oEngineerVisitAssetWorkSheet As EngineerVisitAssetWorkSheet)
                With _database
                    .AddParameter("@EngineerVisitID", oEngineerVisitAssetWorkSheet.EngineerVisitID, True)
                    .AddParameter("@AssetID", oEngineerVisitAssetWorkSheet.AssetID, True)
                    .AddParameter("@FlueTerminationSatisfactoryID", oEngineerVisitAssetWorkSheet.FlueTerminationSatisfactoryID, True)
                    .AddParameter("@FlueFlowTestID", oEngineerVisitAssetWorkSheet.FlueFlowTestID, True)
                    .AddParameter("@SpillageTestID", oEngineerVisitAssetWorkSheet.SpillageTestID, True)
                    .AddParameter("@VentilationProvisionSatisfactoryID", oEngineerVisitAssetWorkSheet.VentilationProvisionSatisfactoryID, True)
                    .AddParameter("@SafetyDeviceOperationID", oEngineerVisitAssetWorkSheet.SafetyDeviceOperationID, True)
                    .AddParameter("@DHWFlowRate", oEngineerVisitAssetWorkSheet.DHWFlowRate, True)
                    .AddParameter("@ColdWaterTemp", oEngineerVisitAssetWorkSheet.ColdWaterTemp, True)
                    .AddParameter("@DHWTemp", oEngineerVisitAssetWorkSheet.DHWTemp, True)
                    .AddParameter("@InletStaticPressure", oEngineerVisitAssetWorkSheet.InletStaticPressure, True)
                    .AddParameter("@InletWorkingPressure", oEngineerVisitAssetWorkSheet.InletWorkingPressure, True)
                    .AddParameter("@MinBurnerPressure", oEngineerVisitAssetWorkSheet.MinBurnerPressure, True)
                    .AddParameter("@MaxBurnerPressure", oEngineerVisitAssetWorkSheet.MaxBurnerPressure, True)
                    .AddParameter("@CO2", oEngineerVisitAssetWorkSheet.CO2, True)
                    .AddParameter("@CO2CORatio", oEngineerVisitAssetWorkSheet.CO2CORatio, True)
                    .AddParameter("@LandlordsApplianceID", oEngineerVisitAssetWorkSheet.LandlordsApplianceID, True)
                    .AddParameter("@ApplianceServiceOrInspectedID", oEngineerVisitAssetWorkSheet.ApplianceServiceOrInspectedID, True)
                    .AddParameter("@ApplianceSafeID", oEngineerVisitAssetWorkSheet.ApplianceSafeID, True)
                    .AddParameter("@VisualConditionOfFlueSatisfactoryID", oEngineerVisitAssetWorkSheet.VisualConditionOfFlueSatisfactoryID, True)
                    .AddParameter("@ApplianceTestedID", oEngineerVisitAssetWorkSheet.ApplianceTestedID, True)
                    .AddParameter("@TankID", oEngineerVisitAssetWorkSheet.TankID, True)
                    .AddParameter("@ReadingType", oEngineerVisitAssetWorkSheet.Reading, True)
                    .AddParameter("@Nozzle", oEngineerVisitAssetWorkSheet.Nozzle, True)
                    .AddParameter("@BurMake", oEngineerVisitAssetWorkSheet.BMake, True)
                    .AddParameter("@BurModel", oEngineerVisitAssetWorkSheet.BModel, True)
                    .AddParameter("@Sweep", oEngineerVisitAssetWorkSheet.SweepOutcomeID, True)
                    .AddParameter("@Overall", oEngineerVisitAssetWorkSheet.OverallID, True)
                    .AddParameter("@Discharge", oEngineerVisitAssetWorkSheet.DischargeID, True)
                End With
            End Sub

            Public Function Products_LastGSRDone(ByVal dateFrom As DateTime, ByVal dateTo As DateTime,
                                                 ByVal regionId As Integer,
                                                 Optional ByVal customerId As Integer = 0,
                                                 Optional ByVal onContract As Integer = 0,
                                                 Optional ByVal tenantsAppliance As Boolean = False,
                                                 Optional ByVal printed As Boolean = False) As DataView
                _database.ClearParameter()
                _database.AddParameter("@DateFrom", dateFrom, True)
                _database.AddParameter("@DateTo", dateTo, True)
                _database.AddParameter("@CustomerID", customerId, True)
                If regionId > 0 Then _database.AddParameter("@RegionID", regionId, True)
                If onContract > 0 Then
                    If onContract = 1 Then
                        _database.AddParameter("@OnContract", True, True)
                    End If
                    If onContract = 2 Then
                        _database.AddParameter("@OnContract", False, True)
                    End If
                End If
                _database.AddParameter("@TenantsAppliance", tenantsAppliance, True)
                _database.AddParameter("@Printed", printed, True)
                Dim dt As DataTable = _database.ExecuteSP_DataTable("Products_LastGSRDone_New")
                dt.TableName = Entity.Sys.Enums.TableNames.tblEngineerVisitAssetWorkSheet.ToString
                Return New DataView(dt)
            End Function

            Public Sub PrintedGSRLettersInsert(ByVal AssetID As Integer, ByVal DateDue As DateTime, Optional ByVal otherContact As Boolean = False)
                _database.ClearParameter()
                _database.AddParameter("@AssetID", AssetID, True)
                _database.AddParameter("@DateDue", DateDue, True)
                _database.AddParameter("PrintedDate", Now, True)
                If otherContact Then
                    _database.AddParameter("OtherContact", True, True)
                End If

                _database.ExecuteSP_NO_Return("PrintedGSRLettersInsert")
            End Sub

#End Region

        End Class

    End Namespace

End Namespace

File: src/Application Specific/Classes/Entity Collections/Jobs/EngineerVisit/EngineerVisitAssetWorkSheet/EngineerVisitAssetWorkSheetSQL.cs
using System;
using System.Data;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace FSM.Entity
{
    namespace EngineerVisitAssetWorkSheets
    {
        public class EngineerVisitAssetWorkSheetSQL
        {
            private Sys.Database _database;

            public EngineerVisitAssetWorkSheetSQL(Sys.Database database)
            {
                _database = database;
            }

            /* TODO ERROR: Skipped RegionDirectiveTrivia */
            public void Delete(int EngineerVisitAssetWorkSheetID)
            {
                _database.ClearParameter();
                _database.AddParameter("@EngineerVisitAssetWorkSheetID", EngineerVisitAssetWorkSheetID, true);
                _database.ExecuteSP_NO_Return("EngineerVisitAssetWorkSheet_Delete");
            }

            public EngineerVisitAssetWorkSheet EngineerVisitAssetWorkSheet_Get(int EngineerVisitAssetWorkSheetID)
            {
                _database.ClearParameter();
                _database.AddParameter("@EngineerVisitAssetWorkSheetID", EngineerVisitAssetWorkSheetID);

                // Get the datatable from the database store in dt
                var dt = _database.ExecuteSP_DataTable("EngineerVisitAssetWorkSheet_Get");
                if (dt is object)
                {
                    if (dt.Rows.Count > 0)
                    {
                        var oEngineerVisitAssetWorkSheet = new EngineerVisitAssetWorkSheet();
                        oEngineerVisitAssetWorkSheet.IgnoreExceptionsOnSetMethods = true;
                        oEngineerVisitAssetWorkSheet.SetEngineerVisitAssetWorkSheetID = dt.Rows[0]["EngineerVisitAssetWorkSheetID"];
                        oEngineerVisitAssetWorkSheet.SetEngineerVisitID = dt.Rows[0]["EngineerVisitID"];
                        oEngineerVisitAssetWorkSheet.SetAssetID = dt.Rows[0]["AssetID"];
                        oEngineerVisitAssetWorkSheet.SetFlueTerminationSatisfactoryID = dt.Rows[0]["FlueTerminationSatisfactoryID"];
                        oEngineerVisitAssetWorkSheet.SetFlueFlowTestID = dt.Rows[0]["FlueFlowTestID"];
                        oEngineerVisitAssetWorkSheet.SetSpillageTestID = dt.Rows[0]["SpillageTestID"];
                        oEngineerVisitAssetWorkSheet.SetVentilationProvisionSatisfactoryID = dt.Rows[0]["VentilationProvisionSatisfactoryID"];
                        oEngineerVisitAssetWorkSheet.SetSafetyDeviceOperationID = dt.Rows[0]["SafetyDeviceOperationID"];
                        oEngineerVisitAssetWorkSheet.SetDHWFlowRate = dt.Rows[0]["DHWFlowRate"];
                        oEngineerVisitAssetWorkSheet.SetColdWaterTemp = dt.Rows[0]["ColdWaterTemp"];
                        oEngineerVisitAssetWorkSheet.SetDHWTemp = dt.Rows[0]["DHWTemp"];
                        oEngineerVisitAssetWorkSheet.SetInletStaticPressure = dt.Rows[0]["InletStaticPressure"];
                        oEngineerVisitAssetWorkSheet.SetInletWorkingPressure = dt.Rows[0]["InletWorkingPressure"];
                        oEngineerVisitAssetWorkSheet.SetMinBurnerPressure = dt.Rows[0]["MinBurnerPressure"];
                        oEngineerVisitAssetWorkSheet.SetMaxBurnerPressure = dt.Rows[0]["MaxBurnerPressure"];
                        oEngineerVisitAssetWorkSheet.SetCO2 = dt.Rows[0]["CO2"];
                        oEngineerVisitAssetWorkSheet.SetCO2CORatio = dt.Rows[0]["CO2CORatio"];
                        oEngineerVisitAssetWorkSheet.SetLandlordsApplianceID = dt.Rows[0]["LandlordsApplianceID"];
                        oEngineerVisitAssetWorkSheet.SetApplianceServiceOrInspectedID = dt.Rows[0]["ApplianceServiceOrInspectedID"];
                        oEngineerVisitAssetWorkSheet.SetApplianceTestedID = dt.Rows[0]["ApplianceTestedID"];
                        oEngineerVisitAssetWorkSheet.SetApplianceSafeID = dt.Rows[0]["ApplianceSafeID"];
                        oEngineerVisitAssetWorkSheet.SetVisualConditionOfFlueSatisfactoryID = dt.Rows[0]["VisualConditionOfFlueSatisfactoryID"];
                        oEngineerVisitAssetWorkSheet.SetDeleted = Conversions.ToBoolean(dt.Rows[0]["Deleted"]);
                        oEngineerVisitAssetWorkSheet.SetNozzle = dt.Rows[0]["Nozzle"];
                        oEngineerVisitAssetWorkSheet.SetReading = dt.Rows[0]["ReadingType"];
                        oEngineerVisitAssetWorkSheet.SetTankID = dt.Rows[0]["TankID"];
                        oEngineerVisitAssetWorkSheet.SetBMake = dt.Rows[0]["BurMake"];
                        oEngineerVisitAssetWorkSheet.SetBModel = dt.Rows[0]["BurModel"];
                        oEngineerVisitAssetWorkSheet.SetSweepOutcomeID = dt.Rows[0]["SweepOutcomeID"];
                        oEngineerVisitAssetWorkSheet.SetOverallID = dt.Rows[0]["OverallID"];
                        oEngineerVisitAssetWorkSheet.SetDischargeID = dt.Rows[0]["DischargeID"];
                        oEngineerVisitAssetWorkSheet.Exists = true;
                        return oEngineerVisitAssetWorkSheet;
                    }
                    else
                    {
                        return null;
                    }
                }
                else
                {
                    return null;
                }
            }

            public DataView EngineerVisitAssetWorkSheet_GetAll()
            {
                _database.ClearParameter();
                var dt = _database.ExecuteSP_DataTable("EngineerVisitAssetWorkSheet_GetAll");
                dt.TableName = Sys.Enums.TableNames.tblEngineerVisitAssetWorkSheet.ToString();
                return new DataView(dt);
            }

            public DataView EngineerVisitAssetWorkSheet_GetForVisit(int EngineerVisitID, int Oil = -1)
            {
                _database.ClearParameter();
                _database.AddParameter("@EngineerVisitID", EngineerVisitID, true);
                _database.AddParameter("@Fuel", Oil, true);
                var dt = _database.ExecuteSP_DataTable("EngineerVisitAssetWorkSheet_GetForVisit");
                dt.TableName = Sys.Enums.TableNames.tblEngineerVisitAssetWorkSheet.ToString();
                return new DataView(dt);
            }

            public DataView Get_Friendly(int engineerVisitAssetWorkSheetId)
            {
                _database.ClearParameter();
                _database.AddParameter("@EngineerVisitAssetWorkSheetID", engineerVisitAssetWorkSheetId, true);
                var dt = _database.ExecuteSP_DataTable("EngineerVisitAssetWorkSheet_Get_Friendly");
                dt.TableName = Sys.Enums.TableNames.tblEngineerVisitAssetWorkSheet.ToString();
                return new DataView(dt);
            }

            public EngineerVisitAssetWorkSheet Insert(EngineerVisitAssetWorkSheet oEngineerVisitAssetWorkSheet)
            {
                _database.ClearParameter();
                AddEngineerVisitAssetWorkSheetParametersToCommand(ref oEngineerVisitAssetWorkSheet);
                oEngineerVisitAssetWorkSheet.SetEngineerVisitAssetWorkSheetID = Sys.Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("EngineerVisitAssetWorkSheet_Insert"));
                oEngineerVisitAssetWorkSheet.Exists = true;
                return oEngineerVisitAssetWorkSheet;
            }

            public DataView EngineerVisitAssetWorkSheet_Search(string criteria)
            {
                _database.ClearParameter();
                _database.AddParameter("@Criteria", criteria, true);
                var dt = _database.ExecuteSP_DataTable("EngineerVisitAssetWorkSheet_Search");
                dt.TableName = Sys.Enums.TableNames.tblEngineerVisitAssetWorkSheet.ToString();
                return new DataView(dt);
            }

            public void Update(EngineerVisitAssetWorkSheet oEngineerVisitAssetWorkSheet)
            {
                _database.ClearParameter();
                _database.AddParameter("@EngineerVisitAssetWorkSheetID", oEngineerVisitAssetWorkSheet.EngineerVisitAssetWorkSheetID, true);
                AddEngineerVisitAssetWorkSheetParametersToCommand(ref oEngineerVisitAssetWorkSheet);
                _database.ExecuteSP_NO_Return("EngineerVisitAssetWorkSheet_Update");
            }

            private void AddEngineerVisitAssetWorkSheetParametersToCommand(ref EngineerVisitAssetWorkSheet oEngineerVisitAssetWorkSheet)
            {
                {
                    var withBlock = _database;
                    withBlock.AddParameter("@EngineerVisitID", oEngineerVisitAssetWorkSheet.EngineerVisitID, true);
                    withBlock.AddParameter("@AssetID", oEngineerVisitAssetWorkSheet.AssetID, true);
                    withBlock.AddParameter("@FlueTerminationSatisfactoryID", oEngineerVisitAssetWorkSheet.FlueTerminationSatisfactoryID, true);
                    withBlock.AddParameter("@FlueFlowTestID", oEngineerVisitAssetWorkSheet.FlueFlowTestID, true);
                    withBlock.AddParameter("@SpillageTestID", oEngineerVisitAssetWorkSheet.SpillageTestID, true);
                    withBlock.AddParameter("@VentilationProvisionSatisfactoryID", oEngineerVisitAssetWorkSheet.VentilationProvisionSatisfactoryID, true);
                    withBlock.AddParameter("@SafetyDeviceOperationID", oEngineerVisitAssetWorkSheet.SafetyDeviceOperationID, true);
                    withBlock.AddParameter("@DHWFlowRate", oEngineerVisitAssetWorkSheet.DHWFlowRate, true);
                    withBlock.AddParameter("@ColdWaterTemp", oEngineerVisitAssetWorkSheet.ColdWaterTemp, true);
                    withBlock.AddParameter("@DHWTemp", oEngineerVisitAssetWorkSheet.DHWTemp, true);
                    withBlock.AddParameter("@InletStaticPressure", oEngineerVisitAssetWorkSheet.InletStaticPressure, true);
                    withBlock.AddParameter("@InletWorkingPressure", oEngineerVisitAssetWorkSheet.InletWorkingPressure, true);
                    withBlock.AddParameter("@MinBurnerPressure", oEngineerVisitAssetWorkSheet.MinBurnerPressure, true);
                    withBlock.AddParameter("@MaxBurnerPressure", oEngineerVisitAssetWorkSheet.MaxBurnerPressure, true);
                    withBlock.AddParameter("@CO2", oEngineerVisitAssetWorkSheet.CO2, true);
                    withBlock.AddParameter("@CO2CORatio", oEngineerVisitAssetWorkSheet.CO2CORatio, true);
                    withBlock.AddParameter("@LandlordsApplianceID", oEngineerVisitAssetWorkSheet.LandlordsApplianceID, true);
                    withBlock.AddParameter("@ApplianceServiceOrInspectedID", oEngineerVisitAssetWorkSheet.ApplianceServiceOrInspectedID, true);
                    withBlock.AddParameter("@ApplianceSafeID", oEngineerVisitAssetWorkSheet.ApplianceSafeID, true);
                    withBlock.AddParameter("@VisualConditionOfFlueSatisfactoryID", oEngineerVisitAssetWorkSheet.VisualConditionOfFlueSatisfactoryID, true);
                    withBlock.AddParameter("@ApplianceTestedID", oEngineerVisitAssetWorkSheet.ApplianceTestedID, true);
                    withBlock.AddParameter("@TankID", oEngineerVisitAssetWorkSheet.TankID, true);
                    withBlock.AddParameter("@ReadingType", oEngineerVisitAssetWorkSheet.Reading, true);
                    withBlock.AddParameter("@Nozzle", oEngineerVisitAssetWorkSheet.Nozzle, true);
                    withBlock.AddParameter("@BurMake", oEngineerVisitAssetWorkSheet.BMake, true);
                    withBlock.AddParameter("@BurModel", oEngineerVisitAssetWorkSheet.BModel, true);
                    withBlock.AddParameter("@Sweep", oEngineerVisitAssetWorkSheet.SweepOutcomeID, true);
                    withBlock.AddParameter("@Overall", oEngineerVisitAssetWorkSheet.OverallID, true);
                    withBlock.AddParameter("@Discharge", oEngineerVisitAssetWorkSheet.DischargeID, true);
                }
            }

            public DataView Products_LastGSRDone(DateTime dateFrom, DateTime dateTo, int regionId, int customerId = 0, int onContract = 0, bool tenantsAppliance = false, bool printed = false)
            {
                _database.ClearParameter();
                _database.AddParameter("@DateFrom", dateFrom, true);
                _database.AddParameter("@DateTo", dateTo, true);
                _database.AddParameter("@CustomerID", customerId, true);
                if (regionId > 0)
                    _database.AddParameter("@RegionID", regionId, true);
                if (onContract > 0)
                {
                    if (onContract == 1)
                    {
                        _database.AddParameter("@OnContract", true, true);
                    }

                    if (onContract == 2)
                    {
                        _database.AddParameter("@OnContract", false, true);
                    }
                }

                _database.AddParameter("@TenantsAppliance", tenantsAppliance, true);
                _database.AddParameter("@Printed", printed, true);
                var dt = _database.ExecuteSP_DataTable("Products_LastGSRDone_New");
                dt.TableName = Sys.Enums.TableNames.tblEngineerVisitAssetWorkSheet.ToString();
                return new DataView(dt);
            }

            public void PrintedGSRLettersInsert(int AssetID, DateTime DateDue, bool otherContact = false)
            {
                _database.ClearParameter();
                _database.AddParameter("@AssetID", AssetID, true);
                _database.AddParameter("@DateDue", DateDue, true);
                _database.AddParameter("PrintedDate", DateAndTime.Now, true);
                if (otherContact)
                {
                    _database.AddParameter("OtherContact", true, true);
                }

                _database.ExecuteSP_NO_Return("PrintedGSRLettersInsert");
            }

            /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        }
    }
}